### PR TITLE
New design, bug fixes

### LINKED ISSUE #5: Make safe portions of structs and slices non-redactable
Today if I make a slice of safe values, or if a struct contains safe values,

unless that struct or slice type has its own `Safeformat` method,

then the entirety of it will be marked as redactable when printed.

Ideally we'd like the recursive struct/slice print logic from the `fmt` package to become redaction-aware.

### LINKED ISSUE #15: %s semantics different from stdlib for nil pointers
I've seen this in a log message, resulting from using %s for printing a nil pointer of a type that implements SafeFormatter on a value receiver.
```
%!s(PANIC=SafeFormatter method: value method github.com/cockroachdb/cockroach/pkg/roachpb.Lease.SafeFormat called using nil *Lease pointer
```

This is different from what fmt.Printf does for the Stringer interface, where it prints <nil>. If possible, it'd certainly take the <nil>. I've found this thread that might describe how to handle this case in the logging lib: golang/go#20995

### PR DESCRIPTION
**Pull request by knz**
Fixes #5 
Fixes #15 

**tldr: this change preserves the existing APIs but makes the library
faster and also able to recursively format structs and arrays.**

Also a new API: `redact.ManualBuffer`, which gives ownership of
swapping between safe and unsafe output to the caller. This is
envisioned to be useful in CockroachDB to help implement `tree.FmtCtx`.

Details:

We are really looking at a major redesign.

The previous design was attempting to preserve Go's own `fmt/print.go`
as much as possible, and was "injecting" special formatting behavior
by wrapping every printed value in a custom object whose type had a
`Format()` method. The idea was to give “ownership” of adding
redaction markers to each argument in turn. That worked for simple
values but there was no good way to make it able to print structs and
arrays recursively.

This new implementation customizes the `print.go` code more heavily.
In particular, the check for safe/unsafe types is now done at
every level of the recursive print. It also makes the code use a
`redact.ManualBuffer` as target during the print evaluation,
which saves string copies (hence a general performance boost).

The new supported cases can be understood by looking at the diffs
on the files `markers_test.go` and `internal/rfmt/registry_test.go`.

Note about the recursive display of struct values: the print logic
uses reflection to print out structs. However, the Go semantics
prevent a package from looking at non-exported fields in structs from
other packages. This means that the print logic is unable to recognize
the various safety interfaces for non-exported struct fields:
`SafeValue`, `SafeFormatter`, `FormatError`, etc. These are thus
considered unsafe regardless of the interfaces implemented.

The safe type registry and the special types `RedactableString` and
`RedactableBytes` are still recognized for non-exported structs
though.

Peformance improvement using the included `bench_test.go`:

```
name       old time/op    new time/op    delta
Redact-32    2.71µs ±22%    1.49µs ± 0%  -45.25%  (p=0.000 n=10+9)

name       old alloc/op   new alloc/op   delta
Redact-32      360B ± 0%      248B ± 0%  -31.11%  (p=0.000 n=10+10)

name       old allocs/op  new allocs/op  delta
Redact-32      12.0 ± 0%       5.0 ± 0%  -58.33%  (p=0.000 n=10+10)
```

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/redact/19)

### CODE REVIEW

**Comment by knz** [comment]
cc @tbg  - long awaited! cost ~3 days of work.

I'm not expecting anyone to be able to review this unless they have already intimate knowledge of Go's `fmt` internals.

Instead:

- I was careful to ensure that the external API of the package has not changed, so we can compare before/after on downstream packages without code changes
- I ensured that the `cockroachdb/errors` library's own test suite passes with this new code without changes. As the `errors` library has a pretty extensive and detailed formatting test suite, that should give us confidence that there are no regressions.
- I'll wait until we re-enable Travis CI and that gives green before merging the branch.
- I'll ensure this is published with a new "major" version (v1.1, we're at v1.0 currently) and clarify there may be unforeseen breakage in the release notes.

**Comment by knz** [comment]
CI passes here. I'm going to check on the errors repo too now.

**Comment by tbg** [comment]
Thank you! Looking forward to all the goodness this brings, 100 plaused
